pkg/odo/genericclioptions/dependencies/dependencies.go

```diff
@@ -0,0 +1,111 @@
+// - complete the Fetch function to instantiate the package's client
+package dependencies
+
+import (
+	"github.com/spf13/cobra"
+
+	"github.com/redhat-developer/odo/pkg/application"
+	_init "github.com/redhat-developer/odo/pkg/init"
+	"github.com/redhat-developer/odo/pkg/init/registry"
+	"github.com/redhat-developer/odo/pkg/kclient"
+	"github.com/redhat-developer/odo/pkg/preference"
+	"github.com/redhat-developer/odo/pkg/testingutil/filesystem"
+)
+
+const (
+	// pkg/application
+	APPLICATION = "DEP_APPLICATION"
+	// pkg/testingutil/filesystem
+	FILESYSTEM = "DEP_FILESYSTEM"
+	// pkg/init
+	INIT = "DEP_INIT"
+	// pkg/preference
+	PREFERENCE = "DEP_PREFERENCE"
+	// pkg/init/registry
+	REGISTRY = "DEP_REGISTRY"
+	// pkg/kclient, can be nil
+	KUBERNETES_NULLABLE = "DEP_KUBERNETES_NULLABLE"
+	// pkg/kclient
+	KUBERNETES = "DEP_KUBERNETES"
+
+	/* Add key for new package here */
+)
+
+// subdeps defines the sub-dependencies
+// Dependencies will be created only once and be reused for sub-dependencies
+var subdeps map[string][]string = map[string][]string{
+	INIT:        {FILESYSTEM, PREFERENCE, REGISTRY},
+	APPLICATION: {KUBERNETES_NULLABLE},
+	/* Add sub-dependencies here, if any */
+}
+
+type Dependency struct {
+	FS                filesystem.Filesystem
+	InitClient        _init.Client
+	PreferenceClient  preference.Client
+	RegistryClient    registry.Client
+	KubernetesClient  kclient.ClientInterface
+	ApplicationClient application.Client
+	/* Add client here */
+}
+
+func Add(command *cobra.Command, dependencies ...string) {
+	if command.Annotations == nil {
+		command.Annotations = map[string]string{}
+	}
+	for _, dependency := range dependencies {
+		_, ok := command.Annotations[dependency]
+		// prevent infinite loop with circular dependencies
+		if !ok {
+			command.Annotations[dependency] = "true"
+			Add(command, subdeps[dependency]...)
+		}
+	}
+}
+
+func isDefined(command *cobra.Command, dependency string) bool {
+	_, ok := command.Annotations[dependency]
+	return ok
+}
+
+func Fetch(command *cobra.Command) (*Dependency, error) {
+	dep := Dependency{}
+	var err error
+	if isDefined(command, PREFERENCE) {
+		dep.PreferenceClient, err = preference.NewClient()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if isDefined(command, FILESYSTEM) {
+		dep.FS = filesystem.DefaultFs{}
+	}
+	if isDefined(command, REGISTRY) {
+		dep.RegistryClient = registry.NewRegistryClient()
+	}
+	if isDefined(command, INIT) {
+		dep.InitClient = _init.NewInitClient(dep.FS, dep.PreferenceClient, dep.RegistryClient)
+	}
+	if isDefined(command, KUBERNETES) || isDefined(command, KUBERNETES_NULLABLE) {
+		dep.KubernetesClient, err = kclient.New()
+		if err != nil && isDefined(command, KUBERNETES) {
+			return nil, err
+		}
+	}
+	if isDefined(command, APPLICATION) {
+		dep.ApplicationClient = application.NewClient(dep.KubernetesClient)
+	}
+
+	/* Instantiate new dependencies here. Take care to instantiate after all sub-dependencies */
```

Comment:
But we aren't instantiating new dependencies here, are we? IIUC, it happens at the beginning of this function. I can't understand what this comment means. 🤔

Reply:
But yes, talking about clients and clientset would be more understandable I agree

Reply:
> I mean "Instantiate new _clients_ here": if a developer creates a new package with a client and wants to add it to this package, he has to instantiate the client in this function

Ah, I just realized that this is more like "to add more stuff, do it here".